### DISCUSSION TITLE
We have "capnp::writeMessageToFd" why is there no "capnp::FdMessageReader" ?

### DISCUSSION THREAD

**Discussion by ThomasArc** [capnproto/capnproto#2032] [General]
There is, of course, capnp::**Stream**FdMessageReader but that doesn't work with UDP.
You have do to the cumbersome flat array stuff if you want to read unpacked UDP.

Am I missing something?
For packed messages, it is nicely orthogonal (and works for UDP):
**capnp::writePackedMessageToFd(fd, builder)** 
**capnp::PackedFdMessageReader**

But for unpacked messages, it is:
**capnp::writeMessageToFd(fd, builder)**
**capnp::StreamFdMessageReader**

Why must unpacked be streamed? Any ideas on this? Missing feature or by design?
I would be very nice to avoid the flat array stuff.

**Comment by kentonv** [comment]
Note of these APIs were ever intended to work with UDP! If any of them do, it's entirely by coincidence!

I suppose `writePackedMessageToFd()` works because it writes the whole message using a single `writev()` call, and so I guess that means it all ends up in a single UDP datagram? (Or fails if the message is too large?) I've never tried this.

On the reading side, reading from UDP is very different than reading from a stream. `StreamFdMessageReader` reads 8 bytes first to find out the size of the segment table, then reads the segment table (if it's larger than 1 entry), then reads all the segments. That's 2-3 separate read calls. It has to do this because it has no idea how large the message is until it has seen the segment table. But this doesn't work with UDP because each read() will read a separate datagram.

When reading from UDP, we have to allocate a buffer big enough for any message we expect to receive, and read into that. That way we receive a whole message in a single call.

`writePackedMesasgeToFd()` and `PackedFdMessageReader` _accidentally_ do this because both of them use buffered I/O. The act of packing requires a buffer, because we don't know in advance how large the data will end up being after it is packed. However, they only "work" if the buffer size happens to be larger than any message you try to send/receive. If you ever happen to have a large message, the sending side would end up fragmenting it into multiple datagrams (which might arrive in any order on the receiving end). Or if you increase the sending-side buffer but the receiving end's buffer is still too small, you'd end up truncating the message, and the reader will end up reading the _next_ datagram and thinking it contains the remainder of the truncated message. Everything will explode.

But if you're OK with all this, then there's a simple way to get the same behavior for unpacked messages: Stack a `StreamFdMessageReader` on top of a `BufferedInputStreamWrapper`, which in turn wraps `FdInputStream`. This adds a layer of buffering, so as long as your messages are smaller than the buffer, it will accidentally work with UDP.

Of course, all of this assumes that on the UDP sending side, you are using connect() to direct the socket to a specific destination IP, but very frequently UDP is used in non-connection-oriented communications where you want to specify a different destination IP for each datagram. We need an entirely different API for that. Similarly, on the receiving end, you can receive packets from any IP on the open socket and you probably want to know who sent each message -- even if you are implementing a two-party stream, you at least want to filter out errant packets from unexpected IPs, right? So the existing APIs are not really appropriate for UDP even in the cases where they accidentally work.

Frankly, I think it probably doesn't make sense for Cap'n Proto to try to provide these APIs. Applications that want to use UDP should probably manage sending and receiving bytes themselves. Cap'n Proto provides perfectly fine APIs to access the raw bytes of a message you have built, or to read the bytes from a buffer you received, so just use those on top of your own network send/recv code.

BTW, when using UDP, I would recommend always allocating your MessageBuilder with a single segment that is the maximum datagram size upfront -- since you have a clear maximum and it's not that big, this makes sense to do. Then, when you send the message, it doesn't have to include the segment table that serialized capnp messages usually start with. You can just send the first segment bytes. Use `builder.getSegmentsForOutput()[0]` to get the bytes on the sending end, and `SegmentArrayMessageReader` to read them on the receiving end. Now you have no need for `capnp/serialize.{h,c++}` at all.

**Comment by ThomasArc** [comment]
Most interesting. Thank you for the reply!

It's kind of cool that it does work for UDP when packed, even though is was not designed for this. I had assumed there was the upper limit of UDP size, and that spanning datagrams would not work.

And yes, to send, I'm using connect() with UDP so there is destination address, but of course it's not a true connection like with TCP. Traffic is local and controlled, and of small size, so it works out really well. No need to discard junk. And no re-ordering is necessary.

For UDP non-packed, I will look at accessing raw bytes as you suggest. I do receive already as raw UDP and then access via flat array. Probably best to do send()/recv() directly, with buffers.  With your explanation, I completely understand why reading UDP directly is not Capn'p functionality.

As for packed working with UDP, is it a _bad_ thing to rely on that? Obviously the datagram size is a limit. Wondering if I should be concerned about future releases breaking this.

And, a question about SOCK_SEQPACKET. I've been using this for local UNIX domain traffic, and it works great - reading/writing packed messages.  Is this one that happens to work too?  Or was that by design?

**Reply by kentonv** [reply]
SOCK_SEQPACKET is also not intended to work. As with UDP, I think you'll see severer problems (hangs, data corruption) if you ever send a message that happens to be larger than the buffer size. The buffer size is an implementation detail so not a great thing to be relying on.

**Comment by ThomasArc** [comment]
This is all very helpful. Thank you, Kenton.
I shall redesign any UDP and SEQQPACKET socket access to use recv() and send() directly, and encode/decode capnp separately.
Definitely best not to rely upon internal implementation details.

For TCP, is that, specifically, _the_ intended use of the Capn'p file descriptor functions?

**Reply by kentonv** [reply]
To be honest, in my experience the synchronous file descriptor functions are not that commonly used. All the code I work on does async network I/O, so the APIs in serialize-async.h are used instead. Sync FDs might be used for file I/O, but often mmap() is preferred there.

But assuming you are OK with blocking the calling thread, then the sync FD APIs should work just fine on TCP sockets, Unix sockets, pipes, and disk files.

**Reply by ThomasArc** [reply]
Yep. All good with blocking the calling thread.
Thank you for all the great info.
